ext/drbbdup/drbbdup.c

```diff
@@ -925,11 +926,22 @@ drbbdup_instrument_dups(void *drcontext, app_pc pc, void *tag, instrlist_t *bb,
         instr_t *end_instr = drbbdup_next_end(next_instr);
         ASSERT(end_instr != NULL, "end instruction cannot be NULL");
 
-        /* Cache first and last instructions. */
-        if (next_instr == end_instr && is_last_special)
+        /* Cache first, first nonlabel and last instructions. */
+        if (next_instr == end_instr && is_last_special) {
             pt->first_instr = last;
-        else
+            pt->first_nonlabel_instr = last;
+        } else {
             pt->first_instr = next_instr; /* Update cache to first instr. */
+            instr_t *first_non_label = next_instr;
+            while (instr_is_label(first_non_label) && first_non_label != end_instr) {
```

Comment:
If it were relatively simple to have the exact instrs passed to the case analysis phase be the ones in the insertion phase too, that would solve multiple problems:

* The client could implement utilities like this is_first_nonlabel itself by recording instr pointers in the case analysis phase.  As is, any time we add some drmgr utility, we might have to go copy it over here, which is not ideal.

* The label problem mentioned in the issue would be naturally solved.

You'd have to cut out each case sequence from the enclosing instrlist, cap it with NULL, and then stitch it back in later.  I'm not sure how much more complex that is than the current cloning of the case instrlist?

Reply:
Just saw this now... I took a different approach - see https://github.com/DynamoRIO/dynamorio/blob/7da96020a382ef8c6516c641dd83f72544e098e2/ext/drbbdup/drbbdup.h#L315

Basically, I provide the entire working instrlist (containing all dups), but the user has to iterate over the basic block using a special API routine called drbbdup_get_next_analysis_instr().

The drbbdup_analyze_case_t call-back is also modified to take the where instruction. Note that we cannot duplicate special instr, like sys calls. (These need to be left at the end - hence the passing of where)

Reply:
I have another PR to handle the case analysis issue and won't be tackling it here. Not marking as resolved just in case you didn't yet read my previous comment.

Reply:
OK, if you have a link to the other PR we can move this discussion there.

> Basically, I provide the entire working instrlist (containing all dups), but the user has to iterate over the basic block using a special API routine called drbbdup_get_next_analysis_instr().

Hmm, requiring a different iterator raises complications: drcachesim has instr_offline_t::identify_elidable_addresses() which walks an instrlist.  It is called both online and offline during post-processing.  The offline decodes from the binary file to create the instrlist and is using DR standalone mode, so it would not be using drbbdup or at least not in a regular setup.  I.e., we have the same code walking an instrlist in an analysis phase and offline.

> The drbbdup_analyze_case_t call-back is also modified to take the where instruction. 

Not sure I follow what "where" means here.